components/brave_news/browser/feed_fetcher.cc

```diff
@@ -152,47 +157,76 @@ void FeedFetcher::OnFetchFeedFetchedFeed(
-  for (const auto& result : results) {
-    total_size += result.items.size();
-  }
-  VLOG(1) << "All feed item fetches done with item count: " << total_size;
-
-  ETags etags;
-  FeedItems feed;
-  feed.reserve(total_size);
-
-  // We want to deduplicate the feed, as the feeds for different regions **may**
-  // have overlap.
-  base::flat_set<GURL> seen;
-
-  for (auto& result : results) {
-    etags[result.key] = result.etag;
-    for (auto& item : result.items) {
-      GURL url;
-      if (item->is_article()) {
-        url = item->get_article()->data->url;
-      } else if (item->is_promoted_article()) {
-        url = item->get_promoted_article()->data->url;
-      }
-
-      // Skip this, we've already seen it.
-      if (!url.is_empty() && seen.contains(url)) {
-        continue;
-      }
-      seen.insert(url);
-
-      feed.push_back(std::move(item));
-    }
-  }
-
-  std::move(callback).Run(std::move(feed), std::move(etags));
+  base::ThreadPool::PostTaskAndReplyWithResult(
+      FROM_HERE,
+      base::BindOnce(
+          [](std::vector<FeedSourceResult> results) {
+            std::size_t total_size = 0;
+            for (const auto& result : results) {
+              total_size += result.items.size();
+            }
+            VLOG(1) << "All feed item fetches done with item count: "
+                    << total_size;
+
+            ETags etags;
+            FeedItems feed;
+            feed.reserve(total_size);
+
+            // We want to deduplicate the feed, as the feeds for different
+            // regions **may** have overlap.
+            std::unordered_set<std::string> seen;
+
+            // reserve |total_size| space in |seen|. This is more than we'll
+            // likely need but should be in the correct ballpark.
+            seen.reserve(total_size);
+
+            for (auto& result : results) {
+              etags[result.key] = result.etag;
+              for (auto& item : result.items) {
+                GURL url;
+                if (item->is_article()) {
+                  url = item->get_article()->data->url;
+                } else if (item->is_promoted_article()) {
+                  url = item->get_promoted_article()->data->url;
+                }
+
+                // Skip this, we've already seen it.
+                auto spec = url.spec();
+                if (!url.is_empty() && seen.contains(spec)) {
+                  continue;
+                }
+                seen.insert(std::move(spec));
+
+                feed.push_back(std::move(item));
+              }
+            }
+
+            return std::make_tuple(std::move(feed), std::move(etags));
+          },
+          std::move(results)),
+      base::BindOnce(
+          [](FetchFeedCallback callback, std::tuple<FeedItems, ETags> result) {
+            std::move(callback).Run(std::move(std::get<0>(result)),
+                                    std::move(std::get<1>(result)));
+          },
+          std::move(callback)));
```

Comment:
The same question about `callback` safety.

Reply:
Yeah, thinking about it, previously the `FeedFetcher` wouldn't fire the callback if it had been destroyed, and we might be relying on that. Nice catch!